{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2bef3c71-e987-4db2-948e-f0c2d78c8d7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "from Bio.Seq import Seq\n",
    "from tqdm import tqdm\n",
    "import re\n",
    "from Bio import Entrez\n",
    "from Bio import SeqIO\n",
    "from Bio import motifs\n",
    "import multiprocessing\n",
    "import warnings\n",
    "\n",
    "def RNA_blast(CDS_name, CDS_seq, taxonomy, que):\n",
    "    try: os.makedirs(f'/home/zhongshitong/data/RNA_seq_result/{taxonomy}/temp/{CDS_name}')\n",
    "    except: pass\n",
    "\n",
    "    os.chdir(f'/home/zhongshitong/data/RNA_seq_result/{taxonomy}/temp/{CDS_name}')\n",
    "\n",
    "    query_sequence = '>'+CDS_name+'\\n'+CDS_seq\n",
    "    query_file = open('RNAseq_queries.fasta','w+')\n",
    "    query_file.write(query_sequence)\n",
    "    query_file.close()\n",
    "    \n",
    "    os.system(f'blastn -query RNAseq_queries.fasta -db /home/zhongshitong/data/RNA_seq_data/{RNAseq_file} -out RNA_blastn_results.txt -evalue 1e-30 -outfmt 6 -max_target_seqs 50000 -num_threads 1')\n",
    "    head = ['qseqid', 'sseqid', 'pident', 'length', 'mismatch', 'gapopen', 'qstart', 'qend', 'sstart', 'send', 'evalue', 'bitscore']\n",
    "    align_result = pd.read_csv('RNA_blastn_results.txt', sep = '\\t|;', engine = 'python', header = None, names = head)\n",
    "    align_result = align_result[align_result['mismatch']==0]\n",
    "    align_result = align_result[align_result['gapopen']==0]\n",
    "    align_result = align_result[align_result['sstart']<align_result['send']]\n",
    "    \n",
    "    align_result = align_result.reset_index(drop=True)\n",
    "    arr = query_sequence.split('\\n')\n",
    "    seq = ''\n",
    "    for i in range(1, len(arr)):\n",
    "        seq = seq + arr[i]\n",
    "    count_arr = []\n",
    "    for i in range(0, len(seq)): \n",
    "        count_arr.append(0)\n",
    "    for i in range(0, len(align_result['sseqid'])):\n",
    "        for j in range(align_result['qstart'][i] -1, align_result['qend'][i]):\n",
    "            count_arr[j] += 1\n",
    "    que.put([count_arr, CDS_name, CDS_seq])\n",
    "\n",
    "def extra_CDSseq(seq_record, CDSseq = '', uplength = 400, downlength = 400):\n",
    "    if 'circular' in seq_record.annotations['topology']:\n",
    "        for feature in seq_record.features:\n",
    "            if feature.type == 'CDS':\n",
    "                if feature.location.strand == 1:\n",
    "                    if (feature.location.end - feature.location.start) == len(seq_record.seq) and len(feature.location.parts) >= 2:\n",
    "                        start_index, end_index = feature.location.parts[0].start - len(seq_record.seq), feature.location.parts[-1].end\n",
    "                    else:\n",
    "                        start_index, end_index = feature.location.start, feature.location.end\n",
    "                    if (start_index - uplength) < 0:\n",
    "                        start_index = start_index + len(seq_record.seq)\n",
    "                        end_index = end_index + len(seq_record.seq)\n",
    "                        CDS_seq = (seq_record.seq+seq_record.seq)[(start_index - uplength):(end_index + downlength)]\n",
    "                    elif (end_index + downlength) >= len(seq_record.seq):\n",
    "                        CDS_seq = (seq_record.seq+seq_record.seq)[(start_index - uplength):(end_index + downlength)]\n",
    "                    else:\n",
    "                        CDS_seq = seq_record.seq[(start_index - uplength):(end_index + downlength)]\n",
    "                else:\n",
    "                    if (feature.location.end - feature.location.start) == len(seq_record.seq) and len(feature.location.parts) >= 2:\n",
    "                        end_index, start_index = feature.location.parts[0].end, feature.location.parts[-1].start - len(seq_record.seq)\n",
    "                    else:\n",
    "                        end_index, start_index = feature.location.end, feature.location.start\n",
    "                    if (end_index + uplength) >= len(seq_record.seq):\n",
    "                        CDS_seq = (seq_record.seq+seq_record.seq)[(start_index - downlength):(end_index + uplength)].reverse_complement()\n",
    "                    elif (start_index - downlength) < 0:\n",
    "                        start_index = start_index + len(seq_record.seq)\n",
    "                        end_index = end_index + len(seq_record.seq)\n",
    "                        CDS_seq = (seq_record.seq+seq_record.seq)[(start_index - downlength):(end_index + uplength)].reverse_complement()\n",
    "                    else:\n",
    "                        CDS_seq = seq_record.seq[(start_index - downlength):(end_index + uplength)].reverse_complement()\n",
    "                try:\n",
    "                    CDSseq += '>' + feature.qualifiers['locus_tag'][0] + '\\n' + str(CDS_seq) + '\\n'\n",
    "                except:\n",
    "                    CDSseq += '>' + feature.qualifiers['protein_id'][0] + '\\n' + str(CDS_seq) + '\\n'\n",
    "    else:\n",
    "        for feature in seq_record.features:\n",
    "            if feature.type == 'CDS':\n",
    "                if feature.location.strand == 1:\n",
    "                    CDS_seq = seq_record.seq[max(0,feature.location.start - uplength):(feature.location.end + downlength)]\n",
    "                else:\n",
    "                    CDS_seq = seq_record.seq[(feature.location.start - downlength):(feature.location.end + uplength)].reverse_complement()\n",
    "                try:\n",
    "                    CDSseq += '>' + feature.qualifiers['locus_tag'][0] + '\\n' + str(CDS_seq) + '\\n'\n",
    "                except:\n",
    "                    CDSseq += '>' + feature.qualifiers['protein_id'][0] + '\\n' + str(CDS_seq) + '\\n'\n",
    "    return CDSseq\n",
    "\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "fb3e84e2-cfa3-4a3c-8a15-fb3fb23137cf",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Program: 100%|██████████████████████████████████████████████████| 3.15k/3.15k [10:08<00:00, 5.17B/s]\n"
     ]
    }
   ],
   "source": [
    "taxonomy, assembly = 'Deinococcus radiodurans', 'GCF_020546685.1'\n",
    "\n",
    "tax_n = taxonomy.replace(' ','_')\n",
    "RNAseq_file = f'{tax_n}_RNAseq.blastdb'\n",
    "\n",
    "seq_data = SeqIO.parse(f'/home/zhongshitong/data/genomes/Deinococcus-Thermus_genomes/data/{assembly}/genomic.gbff', format=\"gb\")\n",
    "\n",
    "try: os.makedirs(f'/home/zhongshitong/data/RNA_seq_result/{taxonomy}')\n",
    "except: pass\n",
    "os.chdir(f'/home/zhongshitong/data/RNA_seq_result/{taxonomy}')\n",
    "\n",
    "CDSseq_file = open('CDSseqs_test.txt', 'w+')\n",
    "CDSseqs = ''\n",
    "for seq_record in seq_data:\n",
    "    CDSseqs += extra_CDSseq(seq_record)\n",
    "\n",
    "CDSseq_file.write(CDSseqs)\n",
    "CDSseq_file.close()\n",
    "\n",
    "CDSseq_data = SeqIO.parse(f'CDSseqs_test.txt', 'fasta')\n",
    "CDSseq_ids = [seq.id for seq in SeqIO.parse(f'CDSseqs_test.txt', 'fasta')]\n",
    "\n",
    "handle = open(f\"/home/zhongshitong/data/up-ORF-motif_simple/assembly/{taxonomy}/meme_out/meme.xml\")\n",
    "record = motifs.parse(handle, \"meme\")\n",
    "data = {'TA_names':[],'TA':[],'AG_names':[],'AG':[]}\n",
    "\n",
    "manager = multiprocessing.Manager()\n",
    "que = manager.Queue()\n",
    "par = 32\n",
    "tot = len(CDSseq_ids)\n",
    "pool = multiprocessing.Pool(par)\n",
    "\n",
    "for seq in CDSseq_data:\n",
    "    CDS_name = seq.id\n",
    "    CDS_seq = str(seq.seq)\n",
    "    pool.apply_async(RNA_blast,(CDS_name, CDS_seq, taxonomy, que))\n",
    "\n",
    "pool.close()\n",
    "\n",
    "re_arr = ''\n",
    "count = 0\n",
    "with tqdm(total = len(CDSseq_ids), desc='Program', leave=True, ncols=100, unit='B', unit_scale=True) as pbar:\n",
    "    while True:\n",
    "        if not que.empty():\n",
    "            result = que.get(True)\n",
    "            count_arr, CDS_name, CDS_seq = result[0], result[1], result[2]\n",
    "            seq_arr = ''\n",
    "            cou_arr = ''\n",
    "            for i in range(len(count_arr)):\n",
    "                seq_arr += CDS_seq[i] + ' '\n",
    "                cou_arr += str(count_arr[i]) + ' '\n",
    "            re_arr += CDS_name+'\\n'+seq_arr+'\\n'+cou_arr+'\\n'\n",
    "            count_arr = count_arr[400:len(count_arr)-400]\n",
    "            pbar.update(1)\n",
    "            count += 1\n",
    "            for motif in record:\n",
    "                if 'TA' in motif.degenerate_consensus:\n",
    "                    id_n = 'TA'\n",
    "                elif 'AG' in motif.degenerate_consensus:\n",
    "                    id_n = 'AG'\n",
    "                found = False\n",
    "                for instance in motif.instances:\n",
    "                    if CDS_name == instance.sequence_name:\n",
    "                        found = True\n",
    "                        break\n",
    "                if found:\n",
    "                    data[id_n].append(sum(count_arr)/len(count_arr))\n",
    "                    data[f'{id_n}_names'].append(CDS_name)\n",
    "            if count == tot:\n",
    "                break\n",
    "        else:\n",
    "            continue\n",
    "            \n",
    "pool.join()\n",
    "os.system(f'rm -rf \"/home/zhongshitong/data/RNA_seq_result/{taxonomy}/temp\"')\n",
    "\n",
    "os.chdir(f'/home/zhongshitong/data/RNA_seq_result/{taxonomy}')\n",
    "re_arr_file = open(f'RNAseq_data_detail.txt', 'w+')\n",
    "re_arr_file.write(re_arr)\n",
    "re_arr_file.close()\n",
    "\n",
    "TA_data = {'TA_names': data['TA_names'], 'TA_data': data['TA'], }\n",
    "pd.DataFrame(TA_data).to_csv(f'TA-RNAseq_data.csv')\n",
    "AG_data = {'AG_names': data['AG_names'], 'AG_data': data['AG'], }\n",
    "pd.DataFrame(AG_data).to_csv(f'AG-RNAseq_data.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "bbb7af33-98a8-4916-bb64-d3914749cb5e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Program: 100%|██████████████████████████████████████████████████| 2.23k/2.23k [13:21<00:00, 2.78B/s]\n"
     ]
    }
   ],
   "source": [
    "taxonomy, assembly = 'Thermus thermophilus', 'GCF_000091545.1'\n",
    "\n",
    "tax_n = taxonomy.replace(' ','_')\n",
    "RNAseq_file = f'{tax_n}_RNAseq_clean.blastdb'\n",
    "\n",
    "seq_data = SeqIO.parse(f'/home/zhongshitong/data/genomes/Deinococcus-Thermus_genomes/data/{assembly}/genomic.gbff', format=\"gb\")\n",
    "\n",
    "try: os.makedirs(f'/home/zhongshitong/data/RNA_seq_result/{taxonomy}')\n",
    "except: pass\n",
    "os.chdir(f'/home/zhongshitong/data/RNA_seq_result/{taxonomy}')\n",
    "\n",
    "CDSseq_file = open('CDSseqs_test.txt', 'w+')\n",
    "CDSseqs = ''\n",
    "for seq_record in seq_data:\n",
    "    CDSseqs += extra_CDSseq(seq_record)\n",
    "\n",
    "CDSseq_file.write(CDSseqs)\n",
    "CDSseq_file.close()\n",
    "\n",
    "CDSseq_data = SeqIO.parse(f'CDSseqs_test.txt', 'fasta')\n",
    "CDSseq_ids = [seq.id for seq in SeqIO.parse(f'CDSseqs_test.txt', 'fasta')]\n",
    "\n",
    "handle = open(f\"/home/zhongshitong/data/up-ORF-motif_simple/assembly/{taxonomy}/meme_out/meme.xml\")\n",
    "record = motifs.parse(handle, \"meme\")\n",
    "data = {'TA_names':[],'TA':[],'AG_names':[],'AG':[]}\n",
    "\n",
    "manager = multiprocessing.Manager()\n",
    "que = manager.Queue()\n",
    "par = 32\n",
    "tot = len(CDSseq_ids)\n",
    "pool = multiprocessing.Pool(par)\n",
    "\n",
    "for seq in CDSseq_data:\n",
    "    CDS_name = seq.id\n",
    "    CDS_seq = str(seq.seq)\n",
    "    pool.apply_async(RNA_blast,(CDS_name, CDS_seq, taxonomy, que))\n",
    "\n",
    "pool.close()\n",
    "\n",
    "re_arr = ''\n",
    "count = 0\n",
    "with tqdm(total = len(CDSseq_ids), desc='Program', leave=True, ncols=100, unit='B', unit_scale=True) as pbar:\n",
    "    while True:\n",
    "        if not que.empty():\n",
    "            result = que.get(True)\n",
    "            count_arr, CDS_name, CDS_seq = result[0], result[1], result[2]\n",
    "            seq_arr = ''\n",
    "            cou_arr = ''\n",
    "            for i in range(len(count_arr)):\n",
    "                seq_arr += CDS_seq[i] + ' '\n",
    "                cou_arr += str(count_arr[i]) + ' '\n",
    "            re_arr += CDS_name+'\\n'+seq_arr+'\\n'+cou_arr+'\\n'\n",
    "            count_arr = count_arr[400:len(count_arr)-400]\n",
    "            pbar.update(1)\n",
    "            count += 1\n",
    "            for motif in record:\n",
    "                if 'TA' in motif.degenerate_consensus:\n",
    "                    id_n = 'TA'\n",
    "                elif 'AG' in motif.degenerate_consensus:\n",
    "                    id_n = 'AG'\n",
    "                found = False\n",
    "                for instance in motif.instances:\n",
    "                    if CDS_name == instance.sequence_name:\n",
    "                        found = True\n",
    "                        break\n",
    "                if found:\n",
    "                    data[id_n].append(sum(count_arr)/len(count_arr))\n",
    "                    data[f'{id_n}_names'].append(CDS_name)\n",
    "            if count == tot:\n",
    "                break\n",
    "        else:\n",
    "            continue\n",
    "            \n",
    "pool.join()\n",
    "os.system(f'rm -rf \"/home/zhongshitong/data/RNA_seq_result/{taxonomy}/temp\"')\n",
    "\n",
    "os.chdir(f'/home/zhongshitong/data/RNA_seq_result/{taxonomy}')\n",
    "re_arr_file = open(f'RNAseq_data_detail.txt', 'w+')\n",
    "re_arr_file.write(re_arr)\n",
    "re_arr_file.close()\n",
    "\n",
    "TA_data = {'TA_names': data['TA_names'], 'TA_data': data['TA'], }\n",
    "pd.DataFrame(TA_data).to_csv(f'TA-RNAseq_data.csv')\n",
    "AG_data = {'AG_names': data['AG_names'], 'AG_data': data['AG'], }\n",
    "pd.DataFrame(AG_data).to_csv(f'AG-RNAseq_data.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb36d7d2",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.20"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
