{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7286385a-8966-443b-917b-38c8f788a1e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import regex as re\n",
    "import pandas as pd\n",
    "import math\n",
    "from Bio import SeqIO\n",
    "from Bio import motifs\n",
    "import warnings\n",
    "from tqdm import tqdm\n",
    "\n",
    "def extra_upORF(seq_record, upORFs = '', uplength = 20, downlength = 0): #提取阅读框上游序列\n",
    "    if uplength + downlength <= 0:\n",
    "        raise ValueError(\"The length of the upORF cannot be zero or negative.\")\n",
    "        return upORFs\n",
    "    if 'circular' in seq_record.annotations['topology']:\n",
    "        for feature in seq_record.features:\n",
    "            if feature.type == 'CDS':\n",
    "                if feature.location.strand == 1:\n",
    "                    if (feature.location.end - feature.location.start) == len(seq_record.seq) and len(feature.location.parts) >= 2:\n",
    "                        start_index = feature.location.parts[0].start\n",
    "                    else:\n",
    "                        start_index = feature.location.start\n",
    "                    if (start_index - uplength) < 0:\n",
    "                        start_index = start_index + len(seq_record.seq)\n",
    "                        upORF_seq = (seq_record.seq+seq_record.seq)[(start_index - uplength):(start_index + downlength)]\n",
    "                    elif (start_index + downlength) >= len(seq_record.seq):\n",
    "                        upORF_seq = (seq_record.seq+seq_record.seq)[(start_index - uplength):(start_index + downlength)]\n",
    "                    else:\n",
    "                        upORF_seq = seq_record.seq[(start_index - uplength):(start_index + downlength)]\n",
    "                else:\n",
    "                    if (feature.location.end - feature.location.start) == len(seq_record.seq) and len(feature.location.parts) >= 2:\n",
    "                        end_index = feature.location.parts[0].end\n",
    "                    else:\n",
    "                        end_index = feature.location.end\n",
    "                    if (end_index + uplength) >= len(seq_record.seq):\n",
    "                        upORF_seq = (seq_record.seq+seq_record.seq)[(end_index - downlength):(end_index + uplength)].reverse_complement()\n",
    "                    elif (end_index - downlength) < 0:\n",
    "                        end_index = end_index + len(seq_record.seq)\n",
    "                        upORF_seq = (seq_record.seq+seq_record.seq)[(end_index - downlength):(end_index + uplength)].reverse_complement()\n",
    "                    else:\n",
    "                        upORF_seq = seq_record.seq[(end_index - downlength):(end_index + uplength)].reverse_complement()\n",
    "                if len(upORF_seq) < (uplength + downlength):\n",
    "                    continue\n",
    "                elif upORF_seq.strip('ATCG') != '':\n",
    "                    continue\n",
    "                else:\n",
    "                    try:\n",
    "                        upORFs += '>' + feature.qualifiers['locus_tag'][0] + '\\n' + str(upORF_seq) + '\\n'\n",
    "                    except:\n",
    "                        upORFs += '>' + feature.qualifiers['protein_id'][0] + '\\n' + str(upORF_seq) + '\\n'\n",
    "        return upORFs\n",
    "    else:\n",
    "        for feature in seq_record.features:\n",
    "            if feature.type == 'CDS':\n",
    "                if feature.location.strand == 1:\n",
    "                    upORF_seq = seq_record.seq[max(0,feature.location.start - uplength):max(0,feature.location.start + downlength)]\n",
    "                else:\n",
    "                    upORF_seq = seq_record.seq[max(0,feature.location.end - downlength):(feature.location.end + uplength)].reverse_complement()\n",
    "                if len(upORF_seq) < (uplength + downlength):\n",
    "                    continue\n",
    "                elif upORF_seq.strip('ATCG') != '':\n",
    "                    continue\n",
    "                else:\n",
    "                    try:\n",
    "                        upORFs += '>' + feature.qualifiers['locus_tag'][0] + '\\n' + str(upORF_seq) + '\\n'\n",
    "                    except:\n",
    "                        upORFs += '>' + feature.qualifiers['protein_id'][0] + '\\n' + str(upORF_seq) + '\\n'\n",
    "        return upORFs\n",
    "\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "try: os.makedirs('/home/zhongshitong/data/up-ORF-motif_simple')\n",
    "except: pass\n",
    "os.chdir('/home/zhongshitong/data/up-ORF-motif_simple')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "30afc271",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ass_name = 'Deinococcus radiodurans'\n",
    "assembly = 'GCF_020546685.1'\n",
    "\n",
    "try: os.makedirs(f'/home/zhongshitong/data/up-ORF-motif_simple/assembly/{ass_name}')\n",
    "except: pass\n",
    "os.chdir(f'/home/zhongshitong/data/up-ORF-motif_simple/assembly/{ass_name}')\n",
    "seq_data = SeqIO.parse(f'/home/zhongshitong/data/genomes/Deinococcus-Thermus_genomes/data/{assembly}/genomic.gbff', format=\"gb\")\n",
    "upORF_file = open('upORFseqs.txt', 'w+')\n",
    "upORFs = ''\n",
    "for seq_record in seq_data:\n",
    "    upORFs += extra_upORF(seq_record)\n",
    "\n",
    "upORF_file.write(upORFs)\n",
    "upORF_file.close()\n",
    "\n",
    "os.system('nohup meme upORFseqs.txt -dna -w 6 -mod zoops -nmotifs 2 -brief 50000 -nostatus -p 1 -nostatus > /dev/null 2>&1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ef29b53b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Program: 100%|███████████████████████████████████████████████████| 3.15k/3.15k [00:30<00:00, 104B/s]\n"
     ]
    }
   ],
   "source": [
    "uplen, downlen = 300, 150\n",
    "os.chdir(f'/home/zhongshitong/data/up-ORF-motif_simple/assembly/{ass_name}')\n",
    "seq_data = SeqIO.parse(f'/home/zhongshitong/data/genomes/Deinococcus-Thermus_genomes/data/{assembly}/genomic.gbff', format=\"gb\")\n",
    "upORF_file = open(f'upORFseqs_{uplen}_{downlen}.txt', 'w+')\n",
    "upORFs = ''\n",
    "for seq_record in seq_data:\n",
    "    upORFs += extra_upORF(seq_record, uplength = uplen, downlength = downlen)\n",
    "\n",
    "upORF_file.write(upORFs)\n",
    "upORF_file.close()\n",
    "\n",
    "seq_res = SeqIO.parse(f'upORFseqs_{uplen}_{downlen}.txt',\"fasta\")\n",
    "seq_ids = [seq.id for seq in SeqIO.parse(f'upORFseqs_{uplen}_{downlen}.txt',\"fasta\")]\n",
    "with tqdm(total = len(seq_ids), desc='Program', leave=True, ncols=100, unit='B', unit_scale=True) as pbar:\n",
    "    columns = ['motif_id', 'motif_alt_id', 'sequence_name', 'start', 'stop', 'strand', 'score', 'p-value', 'q-value', 'matched_sequence']\n",
    "    full_match_result = pd.DataFrame(columns = columns)\n",
    "    os.system(f'nohup fimo --thresh 0.015 --norc meme_out/meme.html upORFseqs_{uplen}_{downlen}.txt > /dev/null 2>&1')\n",
    "    match_result = pd.read_csv(f'fimo_out/fimo.tsv', sep = '\\t|;', engine = 'python', comment='#')\n",
    "    for seq in seq_res:\n",
    "        temp_data = pd.DataFrame(columns = columns)\n",
    "    \n",
    "        temp_upORF = str(seq.seq)\n",
    "        temp_match_result = match_result.loc[(match_result['sequence_name']==str(seq.id))&(match_result['motif_alt_id']=='MEME-1')].reset_index(drop=True)\n",
    "        pbar.update(1)\n",
    "        if temp_match_result.empty:\n",
    "            continue\n",
    "        for j in range(0, len(temp_match_result['motif_id'])):\n",
    "            if temp_match_result['stop'][j] + 20 < uplen: #排除终止密码子的打断\n",
    "                down_seq = temp_upORF[temp_match_result['stop'][j] + 1: uplen]\n",
    "                stopcodes = ['TAG', 'TAA', 'TGA']\n",
    "                td = temp_match_result.loc[[j]]\n",
    "                for stopcode in stopcodes:\n",
    "                    if stopcode in down_seq:\n",
    "                        find_result = re.finditer(stopcode, down_seq, overlapped = True)\n",
    "                        for find_re in find_result:\n",
    "                            if (len(temp_upORF) - re.search(down_seq, temp_upORF).span()[1] + len(down_seq) - find_re.span()[1] - downlen) % 3 == 0:\n",
    "                                td = pd.DataFrame(columns = columns)\n",
    "                                break\n",
    "                if td.empty: continue\n",
    "            down_seq = temp_upORF[temp_match_result['stop'][j] + 1: temp_match_result['stop'][j] + 20]\n",
    "            if 'ATG' in down_seq or 'GTG' in down_seq: #查找下游起始密码子\n",
    "                find_result = re.finditer('[AG]TG', down_seq, overlapped = True)\n",
    "                for find_re in find_result:\n",
    "                    if (len(temp_upORF) - re.search(down_seq, temp_upORF).span()[1] + len(down_seq) - find_re.span()[1] - downlen) % 3 == 0:\n",
    "                        temp_data = pd.concat([temp_data, temp_match_result.loc[[j]]], ignore_index=True)\n",
    "                        break\n",
    "        if temp_data.empty:\n",
    "            continue\n",
    "        else:\n",
    "            best_match_index = 0\n",
    "            startcode_distance = max(uplen, downlen)\n",
    "            temp_data.sort_values(by='start').reset_index(drop=True)\n",
    "            for k in range(0, len(temp_data['motif_id'])):\n",
    "                if k%2 == 0: j = k\n",
    "                else: j = len(temp_data['motif_id']) - k\n",
    "                if abs((temp_data['start'][j] + temp_data['stop'][j])/2 - uplen) < startcode_distance:\n",
    "                    best_match_index = j\n",
    "                    startcode_distance = abs((temp_data['start'][j] + temp_data['stop'][j])/2 - uplen)\n",
    "                if -20 < (temp_data['start'][j] + temp_data['stop'][j])/2 - uplen < 0 or (temp_data['p-value'][j] < 0.006 and 'TA' in temp_data['motif_id'][j]):\n",
    "                    break\n",
    "            full_match_result = pd.concat([full_match_result, temp_data.loc[[best_match_index]]], ignore_index=True)\n",
    "    \n",
    "    for i in range(0, len(full_match_result['motif_id'])):\n",
    "        full_match_result.loc[i, 'start'] = full_match_result['start'][i] - uplen\n",
    "        full_match_result.loc[i, 'stop'] = full_match_result['stop'][i] - uplen\n",
    "    full_match_result.to_csv('full_match_result.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8cb1ba12",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(f'/home/zhongshitong/data/up-ORF-motif_simple/assembly/{ass_name}')\n",
    "seq_data = SeqIO.parse(f'/home/zhongshitong/data/genomes/Deinococcus-Thermus_genomes/data/{assembly}/genomic.gbff', format=\"gb\")\n",
    "CDS_file = open('CDS.txt', 'w+')\n",
    "TA_CDS_file = open('TA_CDS.txt', 'w+')\n",
    "CDS = ''\n",
    "TA_CDS = ''\n",
    "\n",
    "TA_info = [name for name in full_match_result['sequence_name']]\n",
    "\n",
    "for seq_record in seq_data:\n",
    "    for feature in seq_record.features:\n",
    "        if feature.type == 'CDS':\n",
    "            try:\n",
    "                CDS += '>'+feature.qualifiers['locus_tag'][0]+'\\n'+feature.qualifiers['translation'][0]+'\\n'\n",
    "            except:\n",
    "                continue\n",
    "            if feature.qualifiers['locus_tag'][0] in TA_info:\n",
    "                TA_CDS += '>'+feature.qualifiers['locus_tag'][0]+'\\n'+feature.qualifiers['translation'][0]+'\\n'\n",
    "\n",
    "CDS_file.write(CDS)\n",
    "CDS_file.close()\n",
    "TA_CDS_file.write(TA_CDS)\n",
    "TA_CDS_file.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "65fff6e3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Program: 100%|████████████████████████████████████████████████████| 8.00/8.00 [10:22<00:00, 77.8s/B]\n"
     ]
    }
   ],
   "source": [
    "ass_info = {'Calidithermus chliarophilus':'GCF_000430045.1', 'Deinococcus geothermalis': 'GCF_000196275.1', \n",
    "            'Deinobacterium chartae':'GCF_014202645.1', 'Meiothermus silvanus':'GCF_000092125.1', \n",
    "            'Marinithermus hydrothermalis':'GCF_000195335.1', 'Oceanithermus profundus':'GCF_000183745.1', \n",
    "            'Thermus thermophilus':'GCF_000091545.1', 'Truepera radiovictrix':'GCF_000092425.1'}\n",
    "\n",
    "with tqdm(total = len(ass_info), desc='Program', leave=True, ncols=100, unit='B', unit_scale=True) as pbar:\n",
    "    for ass_name in ass_info:\n",
    "        try: os.makedirs(f'/home/zhongshitong/data/up-ORF-motif_simple/assembly/{ass_name}')\n",
    "        except: pass\n",
    "        assembly = ass_info[ass_name]\n",
    "        os.chdir(f'/home/zhongshitong/data/up-ORF-motif_simple/assembly/{ass_name}')\n",
    "        seq_data = SeqIO.parse(f'/home/zhongshitong/data/genomes/Deinococcus-Thermus_genomes/data/{assembly}/genomic.gbff', format=\"gb\")\n",
    "        upORF_file = open('upORFseqs.txt', 'w+')\n",
    "        upORFs = ''\n",
    "        for seq_record in seq_data:\n",
    "            upORFs += extra_upORF(seq_record, uplength = 20)\n",
    "\n",
    "        upORF_file.write(upORFs)\n",
    "        upORF_file.close()\n",
    "\n",
    "        os.system('nohup meme upORFseqs.txt -dna -w 10 -mod zoops -nmotifs 2 -brief 50000 -nostatus -p 1 -nostatus > /dev/null 2>&1')\n",
    "        pbar.update(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd59e2eb",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:R-env]",
   "language": "python",
   "name": "conda-env-R-env-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
